### PR TITLE
Migration of Determinent() from Laplace's method to Gaussian elimination

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**
With help of https://codereview.stackexchange.com/questions/204135/determinant-using-gauss-elimination

Big inner class MatrixElementFraction<T> is necessary because some types like int lose precision when being divided, so that 3 / 2 + 1 / 2 = 1 instead of 2. We compensate it with saying that 3 / 2 + 1 / 2 = (3 + 1) / 2 = 2.

### CODE REVIEW

**Comment by ZacharyPatten** [comment]
This looks really promising. :) 

I still need to review it more but here are some initial thoughts:
1. I still want to keep the old version of the determinent method instead of just deleting it. I would rename it `DeterminentLaplace` and call the new version you have written `DeterminentGaussian` and have the current `Determinent` method just call the `DeterminentGaussian` method.
2. Make sure you check you code before checking in. There appears to be a lot of edits in the Testing project that don't need to happen. This was likely done automatically by your IDE (I assume Visual Studio). I will need any unnecessary changes reverted before I can merge it in.

**Comment by WhiteBlackGoose** [comment]
Actually, I changed the tests and the explanation is the following: unlike Laplace's methods, any "annatural" methods of finding Determinent lose a little bit of precision, so that in one of tests it was 2.00000000001 instead of 2. I will revert them all, rename the method as you just said, and add tests specifically for Gaussian, if you want me to do so.

Though I highly recommend you splitting tests by [TestMethod] so that when you run the tests, you know exactly which case failed.

File: Sources/Towel/Mathematics/Matrix.cs [hunk 1]
```diff
@@ -382,32 +382,176 @@ internal static void GetCofactor(Matrix<T> a, Matrix<T> temp, int p, int q, int
...
+                                Division(Multiplication(fractioned[row, i], fractioned[i, col]), pivotElement));
+                    }
+                }
+			}
+
+			// TODO: should we return zero if determinent's denominator is zero?
+            return determinent.IsDividedByZero ? Constant<T>.Zero : determinent.Value;
+        }
 
-		internal static T GetDeterminant(Matrix<T> a, int n)
```
Comment (WhiteBlackGoose) [inline Sources/Towel/Mathematics/Matrix.cs:387]: I feel like you had wrong number of tabs here

**Comment by WhiteBlackGoose** [comment]
I think I removed all unnecessary code

**Comment by ZacharyPatten** [comment]
I made some edits. Let me know if you disagree with any of them.

There are still more edits I would like to make. The main one is that I already have a `Fraction<T>` class, so if possible I would like to use that rather than a new `MatrixElementFraction<T>` type, but my `Fraction<T>` type is currently automatically reducing after math operations, which would not account for rounding errors like your `MatrixElementFraction<T>` type does. So... I might need to do something like making the automatic reduction in `Fraction<T>` optional... but I'm still trying to think about what the best option is. For now, I'll probably just keep your `MatrixElementFraction<T>` and perhaps we can clean that up sometime in the future.

**Comment by ZacharyPatten** [comment]
I think this PR is in a good state to merge in, but I'd like you opinion of my edits before I do.

**Comment by WhiteBlackGoose** [comment]
They make sense, I didn't do so just because I'm not that familiar with your project. You can merge!

**Comment by ZacharyPatten** [comment]
Thanks for contributing!

**Pull request merged**